### DISCUSSION TITLE
(21) API called with NULL prepared statement?

### DISCUSSION THREAD

**Discussion by Moaneschien** [sqlitebrowser/sqlitebrowser#3598] [Q&A] [answered]
Before  report it as a bug, after running the query below, the error pane says:
```
(21) API called with NULL prepared statement
(21) misuse at line 90775 of [d8cd6d49b4]
```
the result of the query is fine. I've seen it a few times before but didn't pay much attention as the results are ok. This one I caught. Checked it against SQLite 3.45.1 from the command line and it works fine, as expected. Is it a bug?

```
DB Browser for SQLite Version 3.13.99 (Mar 26 2024)
Last commit hash when built: 3e7dd41

Win 11
```
```
CREATE TABLE observationPT10M (
  ds  TEXT NOT NULL,  --'%F'
  pt10m INTEGER NOT NULL, --10 min index, 1 - 144
  energy_consumed Real,
  energy_returned Real,
  UNIQUE(ds,pt10m)
) STRICT;
```
```
CREATE UNIQUE INDEX idx_dspt10m 
    ON observationPT10M (ds, pt10m);
```
```
SELECT ds,
       max(energy_consumed), 
       max(energy_returned)
  FROM observationPT10M
 GROUP BY strftime('%Y', ds)
;
```

**Comment by chrisjlocke** [comment]
Do you get that error each time?

![image](https://github.com/sqlitebrowser/sqlitebrowser/assets/7872218/faae5674-6234-4fe3-81f3-cff69ec9b586)

Run it here with no issues.  Tried both the rc1 of 3.13.1 and the latest nightly.
If you close DB4S and reopen it, do you get the error 'from cold' ?

**Comment by Moaneschien** [comment]
Tried it 'cold' just now, no problem. Yet other queries now give the same error message while the result is fine.

```
  WITH maxvals(ds, maxconsumed, maxreturned) AS (
       SELECT ds,
             max(energy_consumed), 
	     max(energy_returned)
         FROM observationPT10M
        GROUP BY ds
)
SELECT strftime('%F %R', unixepoch(observationPT10M.ds)+((observationPT10M.pt10m/6.0)*3600), 'unixepoch') 
                   AS    ts,
		           energy_consumed - maxconsumed AS consumed_pt10m,
			   energy_returned - maxreturned AS returned_pt10m
		FROM observationPT10M
		JOIN maxvals
			 ON observationPT10M.ds = strftime('%F', maxvals.ds, '+1 day')	--shift one day for substraction
 WHERE observationPT10M.ds 
	      BETWEEN strftime('%F', '2023-06-26', 'weekday 0','-6 days') --monday
       AND strftime('%F', '2023-06-26', 'weekday 0') --sunday
	ORDER BY observationPT10M.ds, observationPT10M.pt10m
;
```

**Comment by chrisjlocke** [comment]
![image](https://github.com/sqlitebrowser/sqlitebrowser/assets/7872218/12657862-22ab-4d56-9872-700b3079e167)

Running that generates no problems.
Next time you get the error, can you take a screenshot, as sometimes this shows more information on what could be generating that.

> API called with NULL prepared statement

An API is commonly used to talk to a remote server.  DB4S doesn't do that (only if you're using the remote tab).  Are you using the remote tab?  Maybe that error is from another action you're performing, and not from running a query.
sqlite is local to DB4S, so for normal queries, there should be no API activity.

**Comment by Moaneschien** [comment]
I'm not using a remote server, `show remote options` is disabled. The SQLite.dll is in de DB Browser directory. 

Image from a cold start running the query.
![21errorAPI](https://github.com/sqlitebrowser/sqlitebrowser/assets/58303834/9a2d5fd0-9eb8-47da-af29-eecf5f506e56)

**Comment by chrisjlocke** [comment, marked as answer]
The screenshot was useful - thanks for posting.
The hash I get (and line number) is different for different versions of DB4S.
Thanks for reporting.  I'll move this to an issue and hopefully someone else can determine whats causing it.
But it can be ignored as it doesn't affect general operations.

**Reply by Moaneschien** [reply]
Thanks Chris, now go and enjoy the easter-eggs ;)
